pkg/daemon/ceph/osd/volume.go

```diff
@@ -613,20 +633,49 @@ func (a *OsdAgent) initializeDevices(context *clusterd.Context, devices *DeviceO
 			"json",
 		}...)
 
-		cvOut, err := context.Executor.ExecuteCommandWithCombinedOutput(baseCommand, reportArgs...)
-		if err != nil {
-			return errors.Wrapf(err, "failed ceph-volume json report: %s", cvOut) // fail return here as validation provided by ceph-volume
-		}
+		cephVersion := a.clusterInfo.CephVersion
 
-		logger.Debugf("ceph-volume report: %+v", cvOut)
+		// ceph version v14.2.13 ~ v14.2.16 changes output of `lvm batch --prepare --report`
+		// use previous logic if ceph version does not fall into this range
+		if !isNewStyledLvmBatch(cephVersion) {
 
-		var cvReport cephVolReport
-		if err = json.Unmarshal([]byte(cvOut), &cvReport); err != nil {
-			return errors.Wrap(err, "failed to unmarshal ceph-volume report json")
-		}
+			cvOut, err := context.Executor.ExecuteCommandWithCombinedOutput(baseCommand, reportArgs...)
+			if err != nil {
+				return errors.Wrapf(err, "failed ceph-volume json report: %s", cvOut) // fail return here as validation provided by ceph-volume
+			}
+
+			logger.Debugf("ceph-volume report: %+v", cvOut)
 
-		if mdPath != cvReport.Vg.Devices {
-			return errors.Errorf("ceph-volume did not use the expected metadataDevice [%s]", mdPath)
+			var cvReport cephVolReport
+			if err = json.Unmarshal([]byte(cvOut), &cvReport); err != nil {
+				return errors.Wrap(err, "failed to unmarshal ceph-volume report json")
+			}
+
+			if mdPath != cvReport.Vg.Devices {
+				return errors.Errorf("ceph-volume did not use the expected metadataDevice [%s]", mdPath)
+			}
+		} else {
+			cvOut, err := context.Executor.ExecuteCommandWithOutput(baseCommand, reportArgs...)
+			if err != nil {
+				return errors.Wrapf(err, "failed ceph-volume json report: %s", cvOut) // fail return here as validation provided by ceph-volume
+			}
+
+			logger.Debugf("ceph-volume reports: %+v", cvOut)
+
+			var cvReports []cephVolReportV2
+			if err = json.Unmarshal([]byte(cvOut), &cvReports); err != nil {
+				return errors.Wrap(err, "failed to unmarshal ceph-volume report json")
+			}
+
+			if len(strings.Split(conf["devices"], " ")) != len(cvReports) {
```

Comment:
Do we really need to check for the length of reports? Just wondering what issues it will really find in the device configuration, or if it might just be a ceph-volume bug?

Reply:
maybe not, I am just wondering if if devices are '/dev/sda1', '/dev/sdb', will ceph volume ignores /dev/sda1, and uses only /dev/sdb?